test/test-tty-duplicate-key.c

```diff
@@ -0,0 +1,313 @@
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == 2);
+  WriteConsoleInputW(handle, &alt_records[1], 1, &written);
+  ASSERT(written == 1);
+
+  /* Emulate transmission of M-a at WSL(#2111) */
+  make_key_event_records(VK_MENU, 0, TRUE, alt_records);
+  WriteConsoleInputW(handle, &alt_records[0], 1, &written);
+  ASSERT(written == 1);
+  make_key_event_records(L'A', LEFT_ALT_PRESSED, TRUE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == 2);
+  WriteConsoleInputW(handle, &alt_records[1], 1, &written);
+  ASSERT(written == 1);
+
+  Sleep(100);
+
+  uv_run(loop, UV_RUN_DEFAULT);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
+TEST_IMPL(tty_composing_character) {
+  int r;
+  int ttyin_fd;
+  uv_tty_t tty_in;
+  uv_loop_t* loop = uv_default_loop();
+  HANDLE handle;
+  INPUT_RECORD records[2];
+  INPUT_RECORD alt_records[2];
+  DWORD written;
+
+  /* Make sure we have an FD that refers to a tty */
+  handle = CreateFileA("conin$",
+                       GENERIC_READ | GENERIC_WRITE,
+                       FILE_SHARE_READ | FILE_SHARE_WRITE,
+                       NULL,
+                       OPEN_EXISTING,
+                       FILE_ATTRIBUTE_NORMAL,
+                       NULL);
+  ASSERT(handle != INVALID_HANDLE_VALUE);
+  ttyin_fd = _open_osfhandle((intptr_t) handle, 0);
+  ASSERT(ttyin_fd >= 0);
+  ASSERT(UV_TTY == uv_guess_handle(ttyin_fd));
+
+  r = uv_tty_init(uv_default_loop(), &tty_in, ttyin_fd, 1);  /* Readable. */
+  ASSERT(r == 0);
+  ASSERT(uv_is_readable((uv_stream_t*) &tty_in));
+  ASSERT(!uv_is_writable((uv_stream_t*) &tty_in));
+
+  r = uv_read_start((uv_stream_t*)&tty_in, tty_alloc, tty_read);
+  ASSERT(r == 0);
+
+  expect_str = EUR_UTF8;
+  expect_nread = strlen(expect_str);
+
+  /* Turn on raw mode. */
+  r = uv_tty_set_mode(&tty_in, UV_TTY_MODE_RAW);
+  ASSERT(r == 0);
+
+  /* Emulate EUR inputs by LEFT ALT+NUMPAD ASCII KeyComos */
+  make_key_event_records(VK_MENU, 0, FALSE, alt_records);
+  alt_records[1].Event.KeyEvent.uChar.UnicodeChar = EUR_UNICODE;
+  WriteConsoleInputW(handle, &alt_records[0], 1, &written);
+  make_key_event_records(VK_NUMPAD0, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  make_key_event_records(VK_NUMPAD1, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  make_key_event_records(VK_NUMPAD2, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  make_key_event_records(VK_NUMPAD8, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  WriteConsoleInputW(handle, &alt_records[1], 1, &written);
+
+  Sleep(100);
```

Comment:
Is it possible to write these tests without `Sleep()`?

Reply:
I removed the `Sleep()` call for the time being. Although there is no problem in the environment at hand, I think there is a possibility that problems will arise in a fast environment.